### PR TITLE
Build only content sets used just for building content, and are not included to ICM

### PR DESCRIPTION
**Pull request by rcerven**
* CLOUDBLD-2046

Signed-off-by: Robert Cerven <rcerven@redhat.com>

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [ ] New feature can be disabled from a configuration file

### CODE REVIEW

**Comment by rcerven** [comment]
tests require https://github.com/containerbuildsystem/osbs-client/pull/992 to be merged

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 1]
```diff
@@ -436,6 +436,10 @@ def __init__(self, data, pulp_data, odcs_config, koji_tag=None, arches=None):
             if data.get(IGNORE_ABSENT_REPOS):
                 self.flags.append(IGNORE_ABSENT_REPOS)
 
+            dev_content_sets = data.get('dev_content_sets', {})
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_resolve_composes.py:439]: Why `dev_content_set`? Shouldn't it be called `build_content_sets`?

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 2]
```diff
@@ -436,6 +436,10 @@ def __init__(self, data, pulp_data, odcs_config, koji_tag=None, arches=None):
...
+            for arch, cont_sets in dev_content_sets.items():
+                self.pulp[arch] = set(cont_sets).union(self.pulp.get(arch, []))
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_resolve_composes.py:441]:
```suggestion
                self.pulp.setdefault(arch, set()).union(cont_sets)
```

File: atomic_reactor/plugins/pre_resolve_composes.py (hunk 1 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_resolve_composes.py:439]: How is case when `dev_content_sets` is `null` handled? (Allowed by schema)

**Comment by MartinBasti** [comment]
This will need also docs update

File: atomic_reactor/plugins/pre_resolve_composes.py (hunk 1 above)
Comment (rcerven) [reply atomic_reactor/plugins/pre_resolve_composes.py:439]: @lcarva @athos-ribeiro @ben-alkov  ?

Comment (rcerven) [reply atomic_reactor/plugins/pre_resolve_composes.py:439]: added handling None

File: atomic_reactor/plugins/pre_resolve_composes.py (hunk 2 above)
Comment (rcerven) [reply atomic_reactor/plugins/pre_resolve_composes.py]: self.pulp is list not set, so this won't work

File: tests/plugins/test_resolve_composes.py [hunk 3]
```diff
@@ -1118,6 +1118,55 @@ def test_only_pulp_repos(self, workflow):
...
 
+    @pytest.mark.parametrize(('content_sets', 'build_only_content_sets'), (
+        (True, True),
+        (True, False),
+        (False, True),
+        (False, False),
+    ))
+    def test_only_content_sets(self, workflow, content_sets, build_only_content_sets):
+        main_cs_list = ['pulp-spam-rpms', 'pulp-bacon-rpms', 'pulp-eggs-rpms']
+        dev_cs_list = ['dev-spam-rpms', 'dev-bacon-rpms', 'dev-eggs-rpms', 'pulp-spam-rpms']
```
Comment (tkdchen) [inline tests/plugins/test_resolve_composes.py:1129]: How about rename it to `build_only_cs_list`?

File: tests/plugins/test_resolve_composes.py [hunk 4]
```diff
@@ -1118,6 +1118,55 @@ def test_only_pulp_repos(self, workflow):
...
+
+        if content_sets:
+            cs_json = {'x86_64': main_cs_list}
+            workflow._tmpdir.join('content_sets.yml').write(yaml.safe_dump(cs_json))
```
Comment (tkdchen) [inline tests/plugins/test_resolve_composes.py:1133]: To make it shorter, it looks variable `cs_json` is useless.

File: tests/plugins/test_resolve_composes.py [hunk 5]
```diff
@@ -1118,6 +1118,55 @@ def test_only_pulp_repos(self, workflow):
...
+
+        containter_json = {'compose': {'pulp_repos': True}}
+        if build_only_content_sets:
+            containter_json['compose']['build_only_content_sets'] = {'x86_64': dev_cs_list}
+        else:
+            containter_json['compose']['build_only_content_sets'] = None
+
+        workflow._tmpdir.join('container.yaml').write(yaml.safe_dump(containter_json))
```
Comment (tkdchen) [inline tests/plugins/test_resolve_composes.py:1141]: Sorry for my nit-pick :) How about:

```python
workflow._tmpdir.join('container.yaml').write(yaml.safe_dump({
    'compose': {
        'pulp_repos': True,
        'build_only_content_sets': {'x86_64': dev_cs_list} if build_only_content_sets else None
    }
})
```

File: tests/plugins/test_resolve_composes.py (hunk 3 above)
Comment (rcerven) [reply tests/plugins/test_resolve_composes.py:1129]: renamed

File: tests/plugins/test_resolve_composes.py (hunk 4 above)
Comment (rcerven) [reply tests/plugins/test_resolve_composes.py:1133]: yes it is just to make it nice :) and also fit in line limit

File: tests/plugins/test_resolve_composes.py (hunk 5 above)
Comment (rcerven) [reply tests/plugins/test_resolve_composes.py:1141]: I'd rather keep simpler formatting as it is now, it is more readable

File: tests/plugins/test_resolve_composes.py [hunk 6]
```diff
@@ -1118,6 +1118,55 @@ def test_only_pulp_repos(self, workflow):
...
+    ))
+    def test_only_content_sets(self, workflow, content_sets, build_only_content_sets):
+        main_cs_list = ['pulp-spam-rpms', 'pulp-bacon-rpms', 'pulp-eggs-rpms']
+        build_only_cs_list = ['dev-spam-rpms', 'dev-bacon-rpms', 'dev-eggs-rpms', 'pulp-spam-rpms']
+
+        if content_sets:
+            cs_json = {'x86_64': main_cs_list}
+            workflow._tmpdir.join('content_sets.yml').write(yaml.safe_dump(cs_json))
+
+        containter_json = {'compose': {'pulp_repos': True}}
```
Comment (ben-alkov) [inline tests/plugins/test_resolve_composes.py:1135]:
```suggestion
        container_json = {'compose': {'pulp_repos': True}}
```

and also a couple of times below here...

Comment (rcerven) [reply tests/plugins/test_resolve_composes.py]: fixed

**Comment by rcerven** [comment]
rn updated: Build only content sets used just for building content, and are not included to ICM

**Pull request merged**